### PR TITLE
rflash for ipmi machines enhancements

### PR DESCRIPTION
**Pull request by gurevichmark**
#5344 

Some `rflash` ipmi enhancements:

1. When displaying message The image file doesn't match this machine also display machine's info
2. If `-d` option is specified return an error if target machine is not "IBM Power S822LC for Big Data"

```
[root@c910f02c05p03 xcat]# rflash fs1 -d /mnt/xcat/iso/supermicro_firmware/p8/OP825/v2.20_BMC1.27
Error: [c910f02c05p03]: fs1: Directory option is not supported for target machine:
 Chassis Type          : Unknown
 Chassis Part Number   : 8335-GTA
 Chassis Serial        : 0000000000000000
 Board Mfg Date        : Sun Dec 31 19:00:00 1995
 Board Mfg             : IBM
 Board Product         : SYSTEM PLANAR
 Board Serial          : 0000000000000000
 Board Part Number     : 0000000000000000
[root@c910f02c05p03 xcat]#
```

```
[root@c910f02c05p03 xcat]# rflash stratton01 -d /mnt/xcat/iso/supermicro_firmware/p8/OP825/v2.20_BMC1.27
Error: [c910f02c05p03]: stratton01: The image file doesn't match target machine:
 Chassis Type          : Unknown
 Chassis Part Number   : 8001-12ABC
 Chassis Serial        : 130124A
 Board Mfg Date        : Sun Dec 31 19:00:00 1995
 Board Mfg             : IBM
 Board Product         : SYSTEM PLANAR
 Board Serial          : OM168S010286
 Board Part Number     : P8DTU
[root@c910f02c05p03 xcat]#
```

### CODE REVIEW

**Comment by whowutwut** [comment]
@gurevichmark  One comment: 

```[root@c910f02c05p03 xcat]# rflash stratton01 -d /mnt/xcat/iso/supermicro_firmware/p8/OP825/v2.20_BMC1.27
Error: [c910f02c05p03]: stratton01: The image file doesn't match target machine:
```

When we get to this point, we are already in the wrong path  and that output may not help us.. because to me, this is a supported machine, the 8001-12C...  But the old code, the hpm_data structure is uninitialized and it's blank...   Is there a way to prevent this from going into this path? 

For example, for the test at 1981, there should be an `else` clause...  

```
2107         $exit_with_success_func->($sessdata->{node}, $callback, "Firmware updated, powering chassis on to populate FRU information...");
2108     }
2109
2110     if (($hpm_data_hash{deviceID} ne $sessdata->{device_id}) ||
2111         ($hpm_data_hash{productID} ne $sessdata->{prod_id}) ||
2112         ($hpm_data_hash{manufactureID} ne $sessdata->{mfg_id})) {
2113         $exit_with_error_func->($sessdata->{node}, $callback,
2114             "The image file doesn't match this machine");
2115     }
2116
```

Essentially, at line 2108, if we dropped down, we should verify that hpm_data_hash is populated, if it's not, then something else went wrong here...  or we could test at 1981 for hpm_data_hash... if it's not populated and it's not one of those, then we missed the support.

**Comment by gurevichmark** [comment]
Second commit added to improve output.

* If big data machine type was missed, like `8001-12C`:
```
[root@c910f02c05p03 xcat]# rflash stratton01 -d /mnt/xcat/iso/supermicro_firmware/p8/OP825/v2.20_BMC1.27
Error: [c910f02c05p03]: stratton01: Flashing of 8001-12C is not supported with pUpdate, supported Model Types: 8001-22C|9006-22C|5104-22C|9006-12C
[root@c910f02c05p03 xcat]#
```

* If user tries to use `-d` for non big data machine:
```
[root@c910f02c05p03 xcat]# rflash fs1 -d /mnt/xcat/iso/supermicro_firmware/p8/OP825/v2.20_BMC1.27
Error: [c910f02c05p03]: fs1: Flashing of 8335-GTA is not supported with pUpdate, supported Model Types: 8001-22C|9006-22C|5104-22C|8001-12C|9006-12C
```

* If somehow could not read data from .hpm file specified:
```
[root@c910f02c05p03 xcat]# rflash fs1 -d /mnt/xcat/iso/supermicro_firmware/p8/OP825/v2.20_BMC1.27
Error: [c910f02c05p03]: fs1: No data was extracted from .hpm update file
[root@c910f02c05p03 xcat]#
```

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 1]
```diff
@@ -2105,13 +2107,30 @@ sub do_firmware_update {
         }
 
         $exit_with_success_func->($sessdata->{node}, $callback, "Firmware updated, powering chassis on to populate FRU information...");
+    } else {
+
+        # The target machine is *NOT* IBM Power S822LC for Big Data (Supermicro)
+        # Only .hpm files is supported for such machine, no directory option is supported
+        if (defined $directory_name) {
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: I think using a regular expression is much easier than split here.

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 2]
```diff
@@ -2105,13 +2107,30 @@ sub do_firmware_update {
...
     }
 
-    if (($hpm_data_hash{deviceID} ne $sessdata->{device_id}) ||
-        ($hpm_data_hash{productID} ne $sessdata->{prod_id}) ||
-        ($hpm_data_hash{manufactureID} ne $sessdata->{mfg_id})) {
-        $exit_with_error_func->($sessdata->{node}, $callback,
-            "The image file doesn't match this machine");
+    # If we fall through here, it is *NOT* IBM Power S822LC for Big Data (Supermicro) and we expact some data in hpm_data_hash.
+    # Verify hpm_data_hash has some values
+    if ((defined $hpm_data_hash{deviceID}) ||
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: 1. Will it be a problem if any of the three attribute(deviceID, productID, manufactureID) is not get from HPM file?
2. I think a correct compare looks like this:
    if ((defined $hpm_data_hash{deviceID} and $hpm_data_hash{deviceID} ne $sessdata->{device_id}) || ...)

**Comment by chuckbrazie** [comment]
I am out of the office until 07/16/2018.

I am out of the office until  Monday July 16.


Note: This is an automated response to your message  "Re: [xcat2/xcat-core]
rflash for ipmi machines enhancements (#5346)" sent on 7/10/18 11:30:40 AM.


This is the only notification you will receive while this person is away.

**Comment by xuweibj** [comment]
Looks good to me. Agree to merge.

**Comment by chuckbrazie** [comment]
I am out of the office until 07/16/2018.

I am out of the office until  Monday July 16.


Note: This is an automated response to your message  "Re: [xcat2/xcat-core]
rflash for ipmi machines enhancements (#5346)" sent on 7/11/18 9:21:11 AM.

This is the only notification you will receive while this person is away.

**Comment by chuckbrazie** [comment]
I am out of the office until 07/16/2018.

I am out of the office until  Monday July 16.


Note: This is an automated response to your message  "Re: [xcat2/xcat-core]
rflash for ipmi machines enhancements (#5346)" sent on 7/11/18 9:21:17 AM.

This is the only notification you will receive while this person is away.

**Pull request merged**